src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs

```diff
@@ -300,37 +300,46 @@ public override void OnException(Activity activity, object payload)
 
         private static string GetUri(HttpRequest request)
         {
-            var builder = new StringBuilder();
-
-            builder.Append(request.Scheme).Append("://");
-
-            if (request.Host.HasValue)
-            {
-                builder.Append(request.Host.Value);
-            }
-            else
+            // this follows the suggestions from https://github.com/dotnet/aspnetcore/issues/28906
+            var scheme = request.Scheme ?? string.Empty;
+
+            // HTTP 1.0 request with NO host header would result in empty Host.
+            // Use placeholder to avoid incorrect URL like "http:///"
+            var host = request.Host.Value ?? UnknownHostName;
+            var pathBase = request.PathBase.Value ?? string.Empty;
+            var path = request.Path.Value ?? string.Empty;
+            var queryString = request.QueryString.Value ?? string.Empty;
+            var length = scheme.Length + Uri.SchemeDelimiter.Length + host.Length + pathBase.Length
+                         + path.Length + queryString.Length;
+#if NETSTANDARD2_1
+            return string.Create(length, (scheme, host, pathBase, path, queryString), (span, parts) =>
             {
-                // HTTP 1.0 request with NO host header would result in empty Host.
-                // Use placeholder to avoid incorrect URL like "http:///"
-                builder.Append(UnknownHostName);
-            }
-
-            if (request.PathBase.HasValue)
-            {
-                builder.Append(request.PathBase.Value);
-            }
-
-            if (request.Path.HasValue)
-            {
-                builder.Append(request.Path.Value);
-            }
-
-            if (request.QueryString.HasValue)
-            {
-                builder.Append(request.QueryString);
-            }
-
-            return builder.ToString();
+                CopyTo(ref span, parts.scheme);
+                CopyTo(ref span, Uri.SchemeDelimiter);
+                CopyTo(ref span, parts.host);
+                CopyTo(ref span, parts.pathBase);
+                CopyTo(ref span, parts.path);
+                CopyTo(ref span, parts.queryString);
+
+                static void CopyTo(ref Span<char> buffer, ReadOnlySpan<char> text)
+                {
+                    if (!text.IsEmpty)
+                    {
+                        text.CopyTo(buffer);
+                        buffer = buffer.Slice(text.Length);
+                    }
+                }
+            });
+#else
+            return new StringBuilder(length)
```

Comment:
Not directly related to this PR - would a ThreadLocal StringBuilder (or something from a pool) help to reduce allocations?

Reply:
ThreadStatic would be one solution, Andrew Lock describes it in his blog: https://andrewlock.net/a-deep-dive-on-stringbuilder-part-5-reducing-allocations-by-caching-stringbuilders-with-stringbuildercache/ and the source for his tests is https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Text/StringBuilderCache.cs